### PR TITLE
Add automatic request retry/backoff when rate limited

### PR DESCRIPTION
**Pull request by GeoffTK**
If the request is rate limited, this library will now sleep for
a bit and then try again. We use an exponential backoff, so the
request should sleep for 1 second, then 2, then 4, 8, 16, and max
out at 32 seconds.

### CODE REVIEW

File: duoapi.go [hunk 1]
```diff
@@ -197,17 +218,27 @@ func (duoapi *DuoApi) Call(method string,
...
-		body, err = ioutil.ReadAll(resp.Body)
-		resp.Body.Close()
+
+	backoffMs := initialBackoffMS
+	for {
+		request, err := http.NewRequest(method, url.String(), nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		resp, err := client.Do(request)
```
Comment (cavemanpi) [inline duoapi.go]: I personally think it would help for code clarity to put the err check here as opposed to inside the if below.

Comment (GeoffTK) [reply duoapi.go]: There is no reason or desire to read the body in the case of a 429 status code, so moving it outside of the if would create extra unnecessary work

File: duo_test.go [hunk 2]
```diff
@@ -154,3 +158,108 @@ func TestNewDuo(t *testing.T) {
...
 	}
 }
+
+func assertRateLimitedCall(t *testing.T, httpResponses []http.Response, finalResponse http.Response, expectedSleepDurations []time.Duration) {
+	httpClient := &mockHttpClient{responses: httpResponses}
+	sleepSvc := &mockSleepService{}
+
+	duo := &DuoApi{
+		ikey: 		"ikey-foo",
+		skey: 		"skey-bar",
```
[the remark this answers is not in the document]
Comment (GeoffTK) [reply duo_test.go]: 👍

File: duoapi.go (hunk 1 above)
Comment (cavemanpi) [reply duoapi.go]: Are you sure you always get a response back even on non-nil err?

https://golang.org/pkg/net/http/#Client.Do
```
On error, any Response can be ignored. A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned Response.Body is already closed.
```

Comment (GeoffTK) [reply duoapi.go]: 👍 great catch

Comment (GeoffTK) [reply duoapi.go]: Let me know if this update addresses that concern.  I added a test in as well for that case

Comment (cavemanpi) [reply duoapi.go]: That's awesome. Thanks!

**Comment by GeoffTK** [comment]
I need to make changes here... I missed a crucial part to this diff

File: duoapi.go [hunk 3]
```diff
@@ -63,8 +72,21 @@ type DuoApi struct {
...
 	host       string
 	userAgent  string
-	apiClient  *http.Client
-	authClient *http.Client
+	apiClient  HttpClient
+	authClient HttpClient
+	sleepSvc   SleepService
+}
+
+type HttpClient interface {
```
Comment (cavemanpi) [inline duoapi.go]: We don't publicly expose the struct fields which make use of these interfaces. Maybe these should be private as well.

**Pull request merged**